Spike: Pass Checkout Session Data Before Authentication

lovo-h opened the pull request:
#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

> The goal of this issue is to explore ways to pass the checkout session data from the merchant store to WooPay before the shopper authenticates.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->
For testing instructions, see 2123-gh-Automattic/woopay.



-------------------

- [ ] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

lovo-h committed:
Ensure WooPayExpressCheckoutButton is cleaned up upon updated_cart_totals

lovo-h committed:
Configure WoopayExpressCheckoutButton to navigate to WooPay checkout

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `7104` or branch name `proto/2093-pass-session-b4-auth` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=proto/2093-pass-session-b4-auth) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 60d34a33bd06794a5afb844afcaf409afd310653
- Build time: 2023-09-11 14:55:35 UTC

_Note: the build is updated when a new commit is pushed to this PR._

lovo-h commented on client/checkout/woopay/express-button/index.js:
```diff
@@ -28,6 +30,14 @@ const renderWooPayExpressCheckoutButton = () => {
+		while ( oldWoopayContainers.length > 0 ) {
+			// Ensure previous buttons are unmounted and cleaned up.
+			const oldWoopayContainer = oldWoopayContainers.pop();
+			ReactDOM.unmountComponentAtNode( oldWoopayContainer );
```
Without this mechanism in place, each time the `updated_cart_totals` event fires, we introduce a new `<WoopayExpressCheckoutButton />` without releasing the previous one(s); leading to a memory leak and preventing the previous `<WoopayExpressCheckoutButton />` from cleaning up.

lovo-h commented on client/checkout/woopay/express-button/woopay-express-checkout-button.js:
```diff
@@ -46,6 +50,20 @@ export const WoopayExpressCheckoutButton = ( {
 	} = useExpressCheckoutProductHandler( api, isProductPage );
 
+	let sessionDataPromise = null;
+	if ( ! isProductPage ) {
```
At the moment, the product page is not supported. Mainly because at this point we are not able to determine whether the item on the product page will be purchased or not. Therefore, in this case, we fall back to the previous approach.

lovo-h replied:
After refactoring, this is no longer an issue. Resolving conversation.

lovo-h commented on client/checkout/woopay/express-button/woopay-express-checkout-button.js:
```diff
@@ -69,7 +87,81 @@ export const WoopayExpressCheckoutButton = ( {
+		};
+
+		const iframe = document.createElement( 'iframe' );
+		iframe.src = getWoopayOtpUrl();
```
Temporarily using the `/otp` URL to establish communication between the merchant site and WooPay site.

lovo-h committed:
Remove addEventListeners() function

Not necessary for just one `addEventListener(...)`.

lovo-h committed:
Only init session data if the WooPay button is pressed

lovo-h committed:
Disable WooPay button while iframe is loading

lovo-h committed:
Replace number with BUTTON_WIDTH_THRESHOLD constant

lovo-h committed:
Prevent HTTP request from being called multiple times

lovo-h committed:
Set useRef default to null

lovo-h committed:
Code cleanup: Improve readability for conditional statement

lovo-h committed:
Disable WooPay button upon click, while loading

lovo-h committed:
Only disable WooPay button onClick if not in preview-mode

lovo-h committed:
Add record WOOPAY_BUTTON_CLICK event

lovo-h committed:
Only preload session data in cart and checkout page, not product page

lovo-h committed:
Add product page support so that onClick, it navigates to WooPay checkout

lovo-h committed:
Prevent multiple iFrames from rendering

Before, the getProductData() and addToCart() functions were causing the
newIframe() callback to be recreated, ultimately causing the useEffect()
that adds the iFrame into the DOM to add another iFrame. Now, the use of
useRef() guarantees we have access to getProductData() and addToCart()
but without causing the newIframe() callback to recreate the iFrame.

lovo-h committed:
Fix: WooPay button from enabling before redirecting to WooPay checkout

This occurs in the product page.

The reason it occurred, was because an `addToCart()` HTTP request is
made, which triggers the `updated_cart_totals` event to fire. This event
then causes a new WooPay button to be rendered. The new button is
enabled by default and so it appears as if the WooPay button gets
enabled right before the browser redirects to the WooPay checkout page.

lovo-h committed:
Add feature flag '_wcpay_feature_woopay_first_party_auth'

lovo-h committed:
Merge branch 'develop' into proto/2093-pass-session-b4-auth

lovo-h committed:
Ensure `isWoopayFirstPartyAuthEnabled` is false when running JS tests

lovo-h committed:
Ensure `fetch()` is defined and returns `json()`

Otherwise, tests that use `fetch()` will fail.

lovo-h committed:
Add changelog

lovo-h committed:
Merge branch 'develop' into proto/2093-pass-session-b4-auth

lovo-h committed:
Remove isDisabled from WooPayExpressCheckoutButton

lovo-h commented:
Closing this PR in favor of the following:
* #7199
